torch/_dynamo/sticky_cache.py

```diff
@@ -0,0 +1,271 @@
+"""
+
+import contextlib
+import dataclasses
+import functools
+import glob
+import logging
+import os
+import pickle
+import shutil
+import types
+from typing import Any, Callable, Dict, Generator, List, Optional, Tuple
+
+import torch
+import torch._inductor.package
+from torch._inductor.codecache import extract_tensor_metadata_for_cache_key
+from torch._inductor.output_code import CompiledAOTI
+from torch._subclasses.fake_tensor import TensorMetadata
+
+from .bytecode_transformation import get_code_keys
+
+
+logger = logging.getLogger(__name__)
+
+
+# Dict type is not hashable, using values instead.
+@functools.lru_cache(maxsize=1)
+def _extract_inputs_metadata_cached(
+    args: Tuple[Any, ...], kwargs: Tuple[Any, ...]
+) -> List[TensorMetadata]:
+    assert all(type(x) == torch.Tensor for x in args)
+    assert all(type(x) == torch.Tensor for x in kwargs)
+    return [extract_tensor_metadata_for_cache_key(t) for t in args + kwargs]
+
+
+def _extract_inputs_metadata(
+    args: Tuple[Any, ...], kwargs: Dict[str, Any]
+) -> List[TensorMetadata]:
+    return _extract_inputs_metadata_cached(args, tuple(kwargs.values()))
+
+
+class _Precompile:
+    """
+    A precompile contains all the serializable information associated with a single
+    compilation in torch.compile(). To restore an execution of compiled code, we will
+    need to serialize the following (not exhaustive):
+      - AOTI compiled code and kernels for forward and backward graph.
+      - AOTAutograd wrappers for things like input mutation.
+      - Dynamo bytecode for mapping Python inputs/outputs.
+      - Dynamo guards for cache keys.
+    """
+
+    def __init__(self) -> None:
+        self.inputs_metadata: Optional[List[TensorMetadata]] = None
+        self.aoti: Optional[CompiledAOTI] = None
+        self.dynamo_code: Optional[types.CodeType] = None
+
+    def add_aoti(self, aoti: CompiledAOTI) -> None:
+        assert self.aoti is None
+        self.aoti = aoti
+
+    def add_inputs_metadata(self, inputs_metadata: List[TensorMetadata]) -> None:
+        assert self.inputs_metadata is None
+        self.inputs_metadata = inputs_metadata
+
+    def add_dynamo_code(self, dynamo_code: types.CodeType) -> None:
+        assert self.dynamo_code is None
+        self.dynamo_code = dynamo_code
+
+    def match_inputs(self, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -> bool:
+        # TODO use dynamo guards when it's ready
+        assert self.inputs_metadata is not None
+        for a, b in zip(self.inputs_metadata, _extract_inputs_metadata(args, kwargs)):
+            if a != b:
+                return False
+        return True
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        assert self.dynamo_code is not None
+        name = next(
+            n for n in self.dynamo_code.co_names if n.startswith("__compiled_fn")
+        )
+        return types.FunctionType(self.dynamo_code, globals={name: self.aoti})(
+            *args, **kwargs
+        )
```

Comment:
This seems wrong.

Dynamo code might depend on other globals in addition to `__compiled_fn0`, both in user code and in generated code.  I think we need to examine the co_names of the code object.

This will also be incorrect if you compile two functions in the same file since the second will be `__compiled_fn1`.

Reply:
I guess an example would be helpful and will be happy to add it to test.

Reply:
You can fullgraph compile two different functions in the same file (or two different shapes for the same functions).

For other globals, put a tensor in global scope and read from it.